DM-32491: Add Butler.get_repo_uri class method for searching repository index

timj opened the pull request:
Pretty straightforward implementation of a repository index.

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj committed:
Add Butler.get_repo_uri class method for searching repository index

ktlim commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -281,6 +281,47 @@ def __init__(self, config: Union[Config, str, None] = None, *,
+        if index_uri is None:
+            raise KeyError("No repository index defined in enviroment variable BUTLER_REPOSITORY_INDEX.")
+
+        config = Config(index_uri)
```
Could it be worth saving this rather than reloading it on every query?  Most of the time there would be exactly one call to this function, but I could see some cases where there might be more than one.  Also, I think it might be useful to provide a method to list all the labels to enable the user to be asked to select one.

timj replied:
`Butler.list_repo_labels()` ?

timj replied:
I've reworked things such that there is now a `ButlerRepoIndex` class that does the heavy lifting and the two `Butler` methods forward directly to that class.  The main issue is whether we bother with the `Butler` class methods or just tell people they have to use `ButlerRepoIndex` directly. The point in favor of `Butler` is that they already have that class imported. The point against is that it's new methods cluttering up the class that are only tangentially related to `Butler`. @mfisherlevine do you have an opinion on:

```python
from lsst.daf.butler import Butler
uri = Butler.get_repo_uri("latiss")
butler = Butler(uri, ...)
```
vs
```python
from lsst.daf.butler import Butler, ButlerRepoIndex
uri = ButlerRepoIndex.get_repo_uri("latiss")
butler = Butler(uri, ...)
```
?

ktlim commented on tests/test_butler.py:
```diff
@@ -482,6 +482,26 @@ def testConstructor(self):
+        # repository.
+        butler_index = Config()
+        butler_index["label"] = self.tmpConfigFile
+        butler_index.dump
```
Um, this looks useless.

timj replied:
indeed. Left over from a code reorganization.

ktlim commented on tests/test_butler.py:
```diff
@@ -482,6 +482,26 @@ def testConstructor(self):
+            with unittest.mock.patch.dict(os.environ, {"BUTLER_REPOSITORY_INDEX": "file://not_found/x.yaml"}):
+                with self.assertRaises(FileNotFoundError):
+                    Butler.get_repo_uri("label")
+            with self.assertRaises(KeyError):
```
This `KeyError` is different from the previous one, though.

timj replied:
You want me to test the contents of the exception?

ktlim replied:
If you wanted to be absolutely safe, then yes, but I think it's OK to treat "missing-label" and "no-labels-defined" as essentially the same.

timj committed:
Add ButlerRepoIndex class for handling index caching and reading.

ktlim commented on python/lsst/daf/butler/_butlerRepoIndex.py:
```diff
@@ -0,0 +1,173 @@
+
+    The index of butler repositories is found by looking for a
+    configuration file at the URI pointed at by the environment
+    variable ``$BUTLER_REPOSITORY_INDEX``. The configuration file
```
Should this have an `LSST_` prefix to help avoid collisions, even if the Butler is being reused outside the project?

timj replied:
Okay. So it looks like butler currently recognizes the following environment variables:

* DAF_BUTLER_PLUGINS
* DAF_BUTLER_CONFIG_PATH
* DAF_BUTLER_CACHE_DIRECTORY
* DAF_BUTLER_CACHE_EXPIRATION_MODE
* DAF_BUTLER_TEST_TMP (for test code)
* LSST_WEBDAV_*. (for ButlerURI webdav implementation)

So you make a strong case for DAF_BUTLER_REPOSITORY_INDEX...

ktlim commented on python/lsst/daf/butler/_butlerRepoIndex.py:
```diff
@@ -0,0 +1,173 @@
+        """
+        index_uri = os.environ.get(cls.index_env_var)
+        if index_uri is None:
+            raise KeyError(f"No repository index defined in enviroment variable {cls.index_env_var}")
```
```suggestion
            raise KeyError(f"No repository index defined in environment variable {cls.index_env_var}")
```

ktlim commented on python/lsst/daf/butler/_butlerRepoIndex.py:
```diff
@@ -0,0 +1,173 @@
+        try:
+            repo_index = cls._read_repository_index_from_environment()
+        except (FileNotFoundError, KeyError):
+            return set()
```
It's not entirely clear to me that it's nicer to return an empty set than to pass along the exception.  It means that there's no available reason for knowing why the set is empty.

timj replied:
I was wondering about that. On the one hand it's nice not to have to put it in a try block and it seems like no results is going to be a strong indicator that something odd is happening. On the other hand you can't then use that knowledge to work out what the actual problem is unless I add logging (or you do `get_repo_uri()` and look at that exception). Maybe some debug log messages wouldn't hurt anyhow.  If you really want it to raise I'll change it.

ktlim replied:
It's a shame you can't return an empty set _and_ an error string :(.  And adding `get_known_repos_error()` doesn't seem reasonable.  I guess just leave it.

ktlim commented on tests/test_butler.py:
```diff
@@ -482,6 +482,35 @@ def testConstructor(self):
+                    self.assertIn("not known to", str(cm.exception))
+        with unittest.mock.patch.dict(os.environ, {"BUTLER_REPOSITORY_INDEX": "file://not_found/x.yaml"}):
+            with self.assertRaises(FileNotFoundError):
+                Butler.get_repo_uri("label")
```
Could easily check for empty set return from `get_known_repos()` here, if you keep that.

timj committed:
Add news fragment

timj committed:
Use DAF_BUTLER_REPOSITORY_INDEX as env var

This is consistent with other butler environment variables
that start with DAF_BUTLER and not BUTLER_.
